### PR TITLE
Expiration policy for first party cookies

### PR DESCRIPTION
**Pull request by sammacbeth**
**Reviewer:** @kdzwinel 

**Depends on:** #540 

## Description:

Sets the `max-age` of cookies set via `document.cookie` from a 3rd party script to 7 days, if they set an expiry longer than 7 days.

Implementation notes:
 * A script is injected in all frames to override the `document.cookie` setter.
 * After injecting this script, we fetch a policy from the extension background, and push it to the injected script. This message also includes the eTLD+1 for the current tab (used to determine if scripts are cross-site).
 * When a cookie is set, we look at the call stack to determine which scripts set the cookie. We compare this to the 1st party domain to determine if the script is 3rd party (this checks same site rather than same origin).
 * If the cookie is set by a 3rd party, and the expiry exceeds the policy threshold, we modify the cookie to set max-age as defined by the policy.

Note this PR currently targets the branch for #540 as it builds on top of those changes.

## Steps to test this PR:
1. Open a website with Google Analytics
2. Check for the `_ga` cookie in the storage tab of devtools
3. The cookie should expire in 2 weeks, rather than 2 years.

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests - Can be added pending deployment of https://github.com/duckduckgo/privacy-test-pages/pull/24

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: integration-test/background/storage.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        const headerCookie = cookies.find(({ name, domain }) => name === 'jsdata' && domain === testPageDomain)
+        expect(headerCookie).toBeTruthy()
+        expect(headerCookie.expires).toBeGreaterThan(Date.now() / 1000)
+    })
+
+    it('blocks 3rd party JS cookies', async () => {
+        const headerCookie = cookies.find(({ name, domain }) => name === 'jsdata' && domain === thirdPartyDomain)
+        expect(headerCookie).toBeUndefined()
+    })
+})
```
Comment (kdzwinel) [inline integration-test/background/storage.js:54]: we need test for the modified expiry, but I understand that is waiting for https://github.com/duckduckgo/privacy-test-pages/pull/24, right?

File: shared/js/content-scripts/block-cookie.js [hunk 2]
```diff
@@ -28,72 +28,109 @@
...
-                        comps[i] = 'max-age=' + nextWeek
-                    }
+    /**
+     * Apply an expiry policy to cookies set via document.cookie.
+     * @param {string} secret Used to detect messages sent from the extension content-script
+     */
+    function applyCookieExpiryPolicy (secret) {
+        const cookieSetter = document.__lookupSetter__('cookie')
+        const cookieGetter = document.__lookupGetter__('cookie')
+        const lineTest = /(\()?(http[^)]+):[0-9]+:[0-9]+(\))?/
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:38]: Can you please check with @jonathanKingston how he is parsing the stack trace? I know he spent quite some time testing different variants in chrome and firefox (https://privacy-test-pages.glitch.me/features/stack-tracing/).
We should probably consolidate both methods at some point.

File: shared/js/content-scripts/block-cookie.js [hunk 3]
```diff
@@ -28,72 +28,109 @@
...
+                    resolve(event.data)
+                    window.removeEventListener('message', messageListener)
                 }
             }
-
-            if (!expiryFound) {
-                comps.push('max-age=' + nextWeek)
+            window.addEventListener('message', messageListener)
+        })
+        console.log('inject cookie test', cookieSetter, secret)
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:50]: we probably want to drop logging before merging or put it behind a flag (there are bunch of console.log's and console.warn's in the file)

File: shared/js/content-scripts/block-cookie.js [hunk 4]
```diff
@@ -28,72 +28,109 @@
...
+                // determine the origins of the scripts in the stack
+                const scriptOrigins = (new Error().stack.split('\n')).reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ tabRegisteredDomain, policy }) => {
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:66]: it looks like tabRegisteredDomain can come undefined if tab is not found. Not sure in what scenarios this can happen, but maybe we want a short path in such case?

File: shared/js/content-scripts/block-cookie.js [hunk 5]
```diff
@@ -28,72 +28,109 @@
...
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:67]: I wonder if this should be `.every` or if we should look only at the last item (one that actually does the setting) or the first item (one that invokes the action) 🤔 Do we know how those stacks look like on various sites? e.g. what happens when some CDN hosted libraries are involved? This might be worth discussing a bit more in tech design.

File: shared/js/content-scripts/block-cookie.js [hunk 6]
```diff
@@ -28,72 +28,109 @@
...
+                    // apply cookie policy
+                    if (expiry > policy.threshold) {
+                        if (maxAgeIdx === -1) {
+                            cookieParts.push(`max-age=${policy.maxAge}`)
+                        } else {
+                            cookieParts.splice(maxAgeIdx, 1, `max-age=${policy.maxAge}`)
+                        }
+                        console.log('update cookie', cookieParts.join(';'), scriptOrigins)
+                        cookieSetter.apply(document, [cookieParts.join(';')])
+                    }
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:93]: gosh, that cookie code is scary. Wish Cookie Storage API (https://wicg.github.io/cookie-store/) was available in FF. But since it isn't… can we somehow unit test this code? 🤔

File: shared/js/content-scripts/block-cookie.js [hunk 7]
```diff
@@ -28,72 +28,109 @@
...
+                })
+            } catch (e) {
+                console.warn('error in cookie override', e)
             }
-
-            const cookieStr = comps.join(';')
-
-            cookieSetter.apply(document, cookieStr)
         })
+        document.__defineGetter__('cookie', cookieGetter)
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:99]: hm, are we doing anything with that getter? Seems like we don't need to touch it at all?

File: shared/js/content-scripts/block-cookie.js [hunk 8]
```diff
@@ -28,72 +28,109 @@
...
+                    const expiresIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('expires'))
+                    if (maxAgeIdx === -1 && expiresIdx === -1) {
+                        // session cookie
+                        return
+                    }
+                    const expiry = maxAgeIdx >= 0
+                        ? parseInt(cookieParts[maxAgeIdx].split('=')[1])
+                        : (new Date(cookieParts[expiresIdx].split('=')[1]) - new Date()) / 1000
+                    // apply cookie policy
+                    if (expiry > policy.threshold) {
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:85]: to be clear, we are modifying the expiry even if page is safelisted, correct? Let's note this in the tech design (I know you proposed it in the comments).

**Review by kdzwinel — COMMENTED** [review]
I left some comments, but overall this looks really solid 💪

File: shared/js/background/chrome-events.es6.js [hunk 9]
```diff
@@ -238,11 +239,21 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
             if (req.checkThirdParty) {
                 const action = {
                     isThirdParty: false,
-                    shouldBlock: false
+                    shouldBlock: false,
+                    tabRegisteredDomain: tldts.parse(sender.tab.url).hostname,
+                    policy: {
+                        threshold: 604800,
```
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:245]: I thought we discussed threshold to be 16d? Let's clarify in tech design.

File: integration-test/background/storage.js (hunk 1 above)
Comment (sammacbeth) [reply integration-test/background/storage.js:54]: Correct.

File: integration-test/background/storage.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        const { browser } = await harness.setup()
+        const page = await browser.newPage()
+        let iframeFullyLoaded = false
+        try {
+            page.on('requestfinished', (req) => {
+                // once we see this url, we can consider the test completed
+                iframeFullyLoaded |= req.url() === `https://${thirdPartyDomain}/privacy-protections/storage-blocking/iframe.js`
+            })
+            await page.goto(`https://${testPageDomain}/privacy-protections/storage-blocking/`, { waitUntil: 'networkidle0' })
+            await page.evaluate('storeData()')
```
[the remark this answers is not in the document]
Comment (sammacbeth) [reply integration-test/background/storage.js:20]: Nice. That should also improve the test as it might catch potential race condition issues.

File: shared/js/content-scripts/block-cookie.js (hunk 2 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:38]: This is the regex copied from his PR.

I'm not sure the best way to share code for this, given this is in a function injected into the page context via `Function.toString()`, so needs to be self contained.

File: shared/js/content-scripts/block-cookie.js (hunk 4 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:66]: It shouldn't ever be undefined - the code that sends it will use the `sender.tab.url` from the `runtime.onMessage` callback if the tab store doesn't have the tab. This case happens when the extension reloads - we don't have tab info for existing open tabs.

File: shared/js/content-scripts/block-cookie.js (hunk 6 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:93]: Again, this code is tricky to extract in a way that we can unittest because of the injection method. However, if we can extract it we could certainly test it by throwing a load of cookie strings see in the wild at it (e.g. from tracker-radar data) to see if there are any missed edge cases.

File: shared/js/content-scripts/block-cookie.js (hunk 7 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:99]: I found if I don't do this, then `document.cookie` is `undefined`.

File: shared/js/content-scripts/block-cookie.js (hunk 8 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:85]: Yes, this policy applies _always_, though with this implementation it would be easy to skip this on pages that the user has allowed.

File: shared/js/content-scripts/block-cookie.js (hunk 2 above)
Comment (jonathanKingston) [reply shared/js/content-scripts/block-cookie.js:39]: Sorry missed this, we could generate a file into the `content-scope` directory from a utils directory perhaps?

File: shared/js/content-scripts/block-cookie.js (hunk 6 above)
Comment (jonathanKingston) [reply shared/js/content-scripts/block-cookie.js:102]: Can we use the `content-scope` injection method I added in my patch? This then keeps the injected code into it's own file which a unit test could load in the same way / import or do whatever without all the other utility code.

File: shared/js/content-scripts/block-cookie.js (hunk 7 above)
Comment (kdzwinel) [reply shared/js/content-scripts/block-cookie.js:118]: TIL!

Comment (kdzwinel) [reply shared/js/content-scripts/block-cookie.js:118]: In fact, can we leave a comment that this is the case? Someone (me in 1mo) may be tempted to remove this thinking it's useless.

File: shared/js/content-scripts/block-cookie.js (hunk 8 above)
Comment (kdzwinel) [reply shared/js/content-scripts/block-cookie.js:101]: Resolving since we discussed it in Tech Design and changed direction to skip safelisted pages for now.

File: shared/js/content-scripts/block-cookie.js [hunk 11]
```diff
@@ -123,12 +110,18 @@
...
+        const scriptString = `{
+            ${imports};
+            (${func.toString()})('${arg}')
+        }`
         const doc = window.wrappedJSObject ? window.wrappedJSObject.document : document
         const scriptElement = doc.createElement('script')
         scriptElement.innerHTML = scriptString
         doc.documentElement.prepend(scriptElement)
+        // remove element immediately so it is not visible to scripts on the page
+        doc.documentElement.removeChild(scriptElement)
```
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:124]: oh, we can do it sync? I thought this would require `setTimeout(,0)` or `script.onload`?

File: unit-test/background/cookie.es6.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        })
+        it('updates existing max-age', () => {
+            const cki = new Cookie('jsdata=; expires=Thu, 01-Jan-1970 00:00:01 GMT; domain=good.third-party.site ;max-age=50')
+            expect(cki.getExpiry()).toBe(50)
+            cki.maxAge = 100
+            expect(cki.getExpiry()).toBe(100)
+            expect(cki.toString().indexOf('max-age=100')).not.toBe(-1)
+        })
+    })
+})
```
Comment (kdzwinel) [inline unit-test/background/cookie.es6.js:48]: Nicely done 👏 Can we throw at it some more "weird but valid" cookie values from https://github.com/nfriedly/set-cookie-parser/blob/master/test/set-cookie-parser.js#L49 ?

File: shared/js/content-scripts/block-cookie.js (hunk 11 above)
Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:124]: I also thought the same, but when I tested it sync also worked.

**Comment by sammacbeth** [comment]
Closing in favour of #570
